### ISSUE TITLE
build: ARM cross compilation issue with 1.5rc1

### ISSUE THREAD

**Issue by tobstarr** [golang/go#12087]
I am getting the error `Illegal instruction` when running a simple program (as simple as a "hello world" with no dependencies, even when only using print("hello world")) on linux/arm (raspberry pi B model) which I cross compiled on linux/amd64 with 1.5.rc1 no matter what value I set GOARM to (tested it with 5, 6 and 7).

Running the program with gdb gives me this error:

    Program received signal SIGILL, Illegal instruction.
    runtime.check () at <GOROOT>/src/runtime/runtime1.go:146

so I guess it is related to the float32 variables here: https://github.com/golang/go/blob/master/src/runtime/runtime1.go#L146

The very same program cross compiles just fine with 1.4.2 (using GOARM 5 and 6) and just gives me the usual error when trying `GOARM=7` 

    runtime: this CPU has no VFPv3 floating point hardware, so it cannot run
    this GOARM=7 binary. Recompile using GOARM=6.

**Comment by ianlancetaylor** [comment]
The release notes ought to document what's going on with ARM before v7.  I'm not sure of the details myself.

**Comment by davecheney** [comment]
Can you show the exact command you used to produce this binary?

Thanks

Dave

**Comment by tobstarr** [comment]
This is how I can reproduce it:

    $ echo $GOROOT
    /home/tobias/go/current

    $ cat main.go
    package main

    func main() {
              print("hello world\n")
    }

    $ go version
    go version go1.5rc1 linux/amd64
    $ GOARM=6 GOARCH=arm go build -o ./ts-pi-cc
    $ file ./ts-pi-cc
    ./ts-pi-cc: ELF 32-bit LSB executable, ARM, EABI5 version 1 (SYSV), statically linked, not stripped
    $ cat ./ts-pi-cc | ssh pi@192.168.178.36 "cat - > ts-pi-cc && chmod a+x ts-pi-cc && ./ts-pi-cc"
    bash: line 1:  1890 Illegal instruction     ./ts-pi-cc

    $ ln -nfs ~/go/go1.4.2 ~/go/current
    $ go version
    go version go1.4.2 linux/amd64
    $ GOARM=6 GOARCH=arm go build -o ./ts-pi-cc
    $ file ./ts-pi-cc
    ./ts-pi-cc: ELF 32-bit LSB executable, ARM, EABI5 version 1 (SYSV), statically linked, not stripped
    $ cat ./ts-pi-cc | ssh pi@192.168.178.36 "cat - > ts-pi-cc && chmod a+x ts-pi-cc && ./ts-pi-cc"
    Hello world

**Comment by davecheney** [comment]
Thanks. I'll try to reproduce this today.

**Comment by mwhudson** [comment]
It's not clear to me from what you say here if the runtime etc is being built with the GOARM setting in effect. Could you try "GOARM=6 GOARCH=arm go build -a -o ./ts-pi-cc" instead?

**Comment by davecheney** [comment]
Hmm, that is true.

Tobias, have you performed any steps to prepare go1.5 for cross
compilation. The tl;Dr is you don't need to and probably shouldn't as of
1.5.

**Comment by davecheney** [comment]
@tobstarr i have not been able to reproduce this issue, so I would like to proceed under the assumption that it is a local issue on your machine.

Can you please do the following.

1. Stop setting GOROOT. In an overwhelming number of these reports, setting GOROOT, no matter how well intentioned, was a contributing factor. 

If you are building Go from source you can follow a procedure like http://dave.cheney.net/2014/09/13/how-to-install-multiple-versions to achieve what you want.

If you are not building from source, and using the tar version of Go, then you can place different versions of Go into something like /usr/local/go1.4, /usr/local/go1.5 and use a symlink, `ln -nfs go1.5 go` to place the version of Go you want in the correct place temporarily, or just move the directories around.

2. Please make sure that your go 1.5 iinstall, /usr/local/go/pkg/ contains no `linux_arm` artifacts. It should not, but if you have made your user the owner of this directory the go tool could have written there (or if you used the sudo hammer).

This advice does _not_ apply to Go 1.4 and earlier, in fact, the opposite is true.

I noticed that in your sample above 
```
$ GOARM=6 GOARCH=arm go build -o ./ts-pi-cc
$ file ./ts-pi-cc
```
Did not print a warning about a missing math.a file. Which, while a bug which i'm going to raise, is an indication that your go1.5 installation has been altered and this may be contributing to the issue.

Thanks

Dave

**Comment by tobstarr** [comment]
I think I found the issue: I guess I used GOARM=7 the first time I cross compiled with my old installation.

It seems that for cross-compilation go creates some object files in `$GOROOT/pkg/linux_arm` but only the first time I cross compile. Could it be that if I set GOARM=7 in the first run those object files are then compiled with GOARM=7. It seems those object files are then also used for all other builds even though I e.g. porvide GOARM=6. 

Deleting `$GOROOT/pkg/linux_arm` actually fixed that issue for me.

Also: I found the error messages/hints in 1.4.2 much better (as they hint to using a different GOARM setting)

**Comment by davecheney** [comment]
> It seems that for cross-compilation go creates some object files in $GOROOT/pkg/linux_arm but only the first time I cross compile. Could it be that if I set GOARM=7 in the first run those object files are then compiled with GOARM=7. It seems those object files are then also used for all other builds even though I e.g. porvide GOARM=6.

This would happen if you used `env GOOS=linux GOARCH=arm GOARM=n go install` _and_ `$GOROOT` was writable by you user.

After Go 1.3 (maybe even in Go 1.3) GOARM stopped being applied at the link stage, and was applied at the compile stage (when the runtime is compiled) which meant the compiled packages in `$GOROOT/pkg` were then customised for the `GOARM` version.

> Deleting $GOROOT/pkg/linux_arm actually fixed that issue for me.

Thanks, closing

> Also: I found the error messages/hints in 1.4.2 much better (as they hint to using a different GOARM setting)

Can you please log a separate issue for this.

**Comment by tobstarr** [comment]
I do not quite understand. Is the recommended solution to delete `GOROOT/pkg/linux_arm` whenever I want to re-compile with a different GOARM? Or is there a way so that those (t seems) GOARM dependent binary files are not kept?

If I am not allowed to write to `GOROOT/pkg` I getting this error: `go install runtime: mkdir /usr/local/go/pkg/linux_arm/: permission denied` that is why I used to install go to `$HOME/go`.

**Comment by davecheney** [comment]
> I do not quite understand. Is the recommended solution to delete GOROOT/pkg/linux_arm whenever I want to re-compile for a different GOARM? Or is there a way so that those (it seems) GOARM dependent binary files are not kept?

There should be no files there, and if Go is trying to write them there, that is a bug. 

When cross compiling you should use `go build` not `go install` to avoid Go complaining it cannot write to `$GOROOT/pkg/...`. If it does write there, because it has permission, then you should delete those files as it is hard to know what `GOARM` value they contain.

> If I am not allowed to write to GOROOT/pkg I get this error: go install runtime: mkdir /usr/local/go/pkg/linux_arm/: permission denied that is why I used to install go to $HOME/go.

You want to avoid Go caching files to `$GOROOT/pkg` because you cannot tell which variant they are.

FWIW, if you do not provide `GOARM` then things will mostly work as `GOARM` defaults to 6 which is the most widely appropriate value.

**Comment by tobstarr** [comment]
Okay, now I understand. Sorry for the confusion. I used `go get` in another script. That is the reason why `pkg/linux_arm` was written in the first place and so those files would also be used even if I ran `go build` in my tests.

I suppose the biggest problem for me was that in 1.4.2 I could use `go get` without any GOARM stickyness and a writeable GOROOT which I can no longer in 1.5. The reason I made GOROOT writeable was actually the necessary cross compilation preparations I needed before 1.5.

I now changed my GOROOT to no longer be writeable by normal users and will always use `go build` for cross compilation. 

So I guess it was actually not a bug but my messed up installation. Thank you for figuring this out!
